gradle.properties

```diff
@@ -14,7 +14,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
-org.gradle.jvmargs=-XX:+UseG1GC -Xmx1g -XX:MaxMetaspaceSize=512m
```

Comment:
I suppose that these jvmargs are also picked up in local runs. If that's the case then the OOM should be reproable locally if we use the same JDK. Having a local repro would make it much easier to test and prove that changes improve the situation. In fact, we could even lower the heap on local runs to see if this is what causes the problem.

Reply:
Yes, these `jvmargs` are used also locally, by setting `-Xmx300m` I get the same error as in CI:

> > Task :linq4j:compileJava
> Note: Some input files use unchecked or unsafe operations.
> Note: Recompile with -Xlint:unchecked for details.
> The Daemon will expire immediately since the JVM garbage collector is thrashing.
> The project memory settings are likely not configured or are configured to an insufficient value.
> The memory settings for this project must be adjusted to avoid this failure.
> These settings can be adjusted by setting 'org.gradle.jvmargs' in 'gradle.properties'.
> The currently configured max heap space is '304 MiB' and the configured max metaspace is '512 MiB'.
> For more information on how to set these values, please refer to https://docs.gradle.org/8.7/userguide/build_environment.html#sec:configuring_jvm_memory in the Gradle documentation.
> To disable this warning, set 'org.gradle.daemon.performance.disable-logging=true'.

So it's pretty much clear that low memory for the JVM can cause this error. What's troubling is that I had a build passing with `-Xmx512m` locally, while we had OOM with `-Xmx1g` in CI...

Tangentially, I have also noticed this warning:

> Configure project :

> The 'sonarqube' task depends on compile tasks. This behavior is now deprecated and will be removed in version 5.x. To avoid implicit compilation, set property 'sonar.gradle.skipCompile' to 'true' and make sure your project is compiled, before analysis has started.
> 
> The 'sonar' task depends on compile tasks. This behavior is now deprecated and will be removed in version 5.x. To avoid implicit compilation, set property 'sonar.gradle.skipCompile' to 'true' and make sure your project is compiled, before analysis has started.

I am wondering if setting the suggested property, and building and running sonar in two separate Gradle commands would help by spawning a fresh single-use daemon:

`./gradlew build --no-parallel --no-daemon` (or even `./gradlew build` as it's separate from Sonar?)

and 

`./gradlew --no-parallel --no-daemon jacocoAggregateTestReport sonar -PenableJacoco -Porg.sonarqube.version=4.4.1.3373 -Dsonar.branch.name=${BRANCH_NAME} -Dsonar.login=${SONAR_TOKEN}`

If all is good, subsequent tasks should not interfere with each others, so ideally there should not be any difference, but in practice the situation might be more complicated (my knowledge of Sonar is limited).

WDYT?

Reply:
I also tested it many times, and fortunately I did not trigger OOM

Reply:
https://ci-builds.apache.org/job/Calcite/job/Calcite-sonar/job/PR-3864/14/cloudbees-pipeline-explorer/ failed again with OOM with `org.gradle.jvmargs=-XX:+UseG1GC -Xmx768m -XX:MaxMetaspaceSize=512m`, will retrigger another time with these same parameters to see if the failure is confirmed.

If that's the case, both locally and in CI we have found that the same transient OOM error we generally see depends on these parameters, and hopefully a 50% increase can stabilize CI.